staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/fairqueuing.go

```diff
@@ -0,0 +1,623 @@
+	if p.GetQueue().GetIndex() >= qs.desiredNumQueues &&
+		len(p.GetQueue().GetPackets()) == 0 &&
+		p.GetQueue().GetRequestsExecuting() == 0 {
+		qs.Queues = removeQueueAndUpdateIndexes(qs.Queues, p.GetQueue().GetIndex())
+		// At this point, if the qs is quiescing,
+		// has zero requests executing, and has zero requests enqueued
+		// then a call to the EmptyHandler should be forked.
+		if qs.Quiescent && qs.NumPacketsEnqueued == 0 &&
+			qs.GetRequestsExecuting() == 0 {
+			// then a call to the EmptyHandler should be forked.
+			go func() {
+				// TODO(aaron-prindle) store the emptyHandler to call it here?
+			}()
+		}
+	}
+}
+
+// dequeue dequeues a packet from the QueueSetImpl
+func (qs *QueueSetImpl) dequeue() (FQPacket, bool) {
+
+	queue := qs.selectQueue()
+
+	if queue == nil {
+		return nil, false
+	}
+	packet, ok := queue.Dequeue()
+
+	if ok {
+		// When a request is dequeued for service -> qs.VirStart += G
+		virstart := queue.GetVirStart()
+		virstart += qs.estimatedServiceTime
+		queue.SetVirStart(virstart)
+
+		packet.SetStartTime(qs.clock.Now())
+		// request dequeued, service has started
+		queue.SetRequestsExecuting(queue.GetRequestsExecuting() + 1)
+	} else {
+		// TODO(aaron-prindle) verify this statement is needed...
+		return nil, false
+	}
+	qs.NumPacketsEnqueued--
+	return packet, ok
+}
+
+// Dequeue dequeues a packet from the QueueSetImpl
+func (qs *QueueSetImpl) Dequeue() (FQPacket, bool) {
+	qs.LockAndSyncTime()
+	defer qs.Lock.Unlock()
+	return qs.dequeue()
+}
+
+// isEmpty is a convenience method that returns 'true' when all of the queues
+// in an QueueSetImpl have no packets (and is "empty")
+func (qs *QueueSetImpl) isEmpty() bool {
+	return qs.NumPacketsEnqueued == 0
+}
+
+// DequeueWithChannelAsMuchAsPossible runs a loop, as long as there
+// are non-empty queues and the number currently executing is less than the
+// assured concurrency value.  The body of the loop uses the fair queuing
+// technique to pick a queue, dequeue the request at the head of that
+// queue, increment the count of the number executing, and send `{true,
+// handleCompletion(that dequeued request)}` to the request's channel.
+func (qs *QueueSetImpl) DequeueWithChannelAsMuchAsPossible() {
+	for !qs.isEmpty() && qs.GetRequestsExecuting() < qs.concurrencyLimit {
+		_, ok := qs.dequeueWithChannel()
+		// TODO(aaron-prindle) verify checking ok makes senes
+		if !ok {
+			break
+		}
+	}
+}
+
+// dequeueWithChannel is convenience method for dequeueing packets that
+// require a message to be sent through the packets channel
+// this is a required pattern for the QueueSetSystem the QueueSetImpl supports
+func (qs *QueueSetImpl) dequeueWithChannel() (FQPacket, bool) {
+	packet, ok := qs.dequeue()
+	if !ok {
+		return nil, false
+	}
+	reqMgmtPkt, conversionOK := packet.(*Packet)
+	if !conversionOK {
+		// TODO(aaron-prindle) log an error
+		return nil, false
+	}
+	reqMgmtPkt.DequeueChannel <- true
```

Comment:
To support testing with a FakeEventClock, the QueueSet will need not only a clock but also a `*sync.WaitGroup`, on which `Add(1)` will need to be called here to account for the unblocking of whatever goroutine receives the value sent here.  Same for sends of `false`.

Reply:
updated

Reply:
Sorry, I forgot to mention that the `*sync.WaitGroup` can be `nil`, and the Add gets called only if the pointer is non-nil.  Same everywhere that uses the pointer.

Reply:
updated, including nil check.